compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3216,6 +3215,252 @@ TR::Register *OMR::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::Co
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   TR::LabelSymbol* loopLabel = generateLabelSymbol(cg);
+
+   generateLabelInstruction(LABEL, node, startLabel, cg);
+
+   // Example:
+   // ; obtain distance from source to destination
+   //   SUB   EDI,ESI
+   //
+   // ; decide direction
+   //   CMP   EDI,ECX
+   //   LEA   EDI,[EDI+ESI]
+   //   LEA   EAX,[ESI+ECX-8]
+   //   CMOVB ESI,EAX
+   //   SETAE AL
+   //   MOVZX EAX,AL
+   //   LEA   EAX,[2*EAX-1]
+   //
+   // ; copy
+   //   SHR  ECX,3
+   //   JRCXZ endLabel
+   // loopLabel:
+   //   MOVQ XMM0,[ESI]
+   //   MOVQ [EDI+ESI],XMM0
+   //   LEA  ESI,[ESI+8*EAX]
+   //   LOOP loopLabel
+   // endLable:
+   //   # LOOP END
+   generateRegRegInstruction(SUBRegReg(), node, dstReg, srcReg, cg);
+   if (node->isForwardArrayCopy())
+      {
+      generateRegImmInstruction(MOVRegImm4(), node, scrach, 1, cg);
+      }
+   else // decide direction during runtime
+      {
+      generateRegRegInstruction(CMPRegReg(), node, dstReg, sizeReg, cg);
+      generateRegMemInstruction(LEARegMem(), node, scrach, generateX86MemoryReference(srcReg, sizeReg, 0, -8, cg), cg);
+      generateRegRegInstruction(CMOVBRegReg(), node, srcReg, scrach, cg);
+      generateRegInstruction(SETAE1Reg, node, scrach, cg);
+      generateRegRegInstruction(MOVZXReg4Reg1, node, scrach, scrach, cg);
+      generateRegMemInstruction(LEARegMem(), node, scrach, generateX86MemoryReference(NULL, scrach, 1, -1, cg), cg);
+      }
+
+   generateRegImmInstruction(SHRRegImm1(), node, sizeReg, 3, cg);
+   generateLabelInstruction(JRCXZ1, node, endLabel, cg);
+   generateLabelInstruction(LABEL, node, loopLabel, cg);
+   generateRegMemInstruction(MOVQRegMem, node, XMM, generateX86MemoryReference(srcReg, 0, cg), cg);
+   generateMemRegInstruction(MOVQMemReg, node, generateX86MemoryReference(dstReg, srcReg, 0, cg), XMM, cg);
+   generateRegMemInstruction(LEARegMem(), node, srcReg, generateX86MemoryReference(srcReg, scrach, 3, cg), cg);
+   generateLabelInstruction(LOOP1, node, loopLabel, cg);
+
+   generateLabelInstruction(LABEL, node, endLabel, dependencies, cg);
+   cg->stopUsingRegister(XMM);
+   cg->stopUsingRegister(scrach);
+   }
+
+static void arrayCopyDefault(TR::Node* node, uint8_t elementSize, TR::Register* dstReg, TR::Register* srcReg, TR::Register* sizeReg, TR::CodeGenerator* cg)
+   {
+   TR::RegisterDependencyConditions* dependencies = generateRegisterDependencyConditions((uint8_t)3, (uint8_t)3, cg);
+   dependencies->addPreCondition(srcReg, TR::RealRegister::esi, cg);
+   dependencies->addPreCondition(dstReg, TR::RealRegister::edi, cg);
+   dependencies->addPreCondition(sizeReg, TR::RealRegister::ecx, cg);
+   dependencies->addPostCondition(srcReg, TR::RealRegister::esi, cg);
+   dependencies->addPostCondition(dstReg, TR::RealRegister::edi, cg);
+   dependencies->addPostCondition(sizeReg, TR::RealRegister::ecx, cg);
+
+   TR_X86OpCodes REPMOVS;
```

Comment:
I find the use of the capitalized auto REPMOVS confusing as its not a real opcode (though its used in the context of real opcodes) and inconsistent.  Please use lower-case as per convention throughout the code generator.